File: 2019/3/solution.js
const fs = require('fs');
const path = require('path');

const input = fs.readFileSync(path.join(__dirname, 'input.txt'), 'utf-8', (err, data) => {
  if (err) {
    console.log(err)
  } else {
    return data;
  }
}).split('\n').map((a) => a.split(','));

const pos2String = (pos) => pos[0] + ',' + pos[1];

class Wire {
  constructor(input) {
    this.input = input;
    this.set = new Set();
    this.steps = new Map();
    this.mapWire();
  }

  mapWire() {
    const pos = [0, 0];
    let step = 0;
    for (const part of this.input) {
      const direction = part[0];
      const distance = Number(part.slice(1));
      const increment = [0, 0];
      switch (direction) {
        case 'U':
          increment[1] = -1;
          break;
        case 'D':
          increment[1] = 1;
          break;
        case 'L':
          increment[0] = -1;
          break;
        case 'R':
          increment[0] = 1;
      }
      for (let i = 0; i < distance; i++) {
        pos[0] += increment[0];
        pos[1] += increment[1];
        step++;
        const posString = pos2String(pos);
        this.set.add(posString);
        if (!this.steps.has(posString)) {
          this.steps.set(posString, step);
        }
      }
    }
  }
}

const wire1 = new Wire(input[0]);
const wire2 = new Wire(input[1]);

const intersections = wire1.set.intersection(wire2.set);

let part1 = Infinity;
let part2 = Infinity;

for (const pos of intersections) {
  const manhattanDistance = pos.split(',').map(Number).map(Math.abs).reduce((a, b) => a + b);
  const combinedSteps = wire1.steps.get(pos) + wire2.steps.get(pos);
  part1 = Math.min(part1, manhattanDistance);
  part2 = Math.min(part2, combinedSteps);
}

console.log(part1);
console.log(part2);

File: 2019/3/solution.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(solve(puzzleInput))
}

type Segment struct {
	direction string
	distance  int
}

var directions = map[string][2]int{"U": {0, 1}, "D": {0, -1}, "L": {-1, 0}, "R": {1, 0}}

func parseInput(data []string) [2][]Segment {
	var wires [2][]Segment
	for i, s := range data {
		parts := strings.Split(s, ",")
		var wire []Segment
		for _, part := range parts {
			n, err := strconv.Atoi(part[1:])
			if err != nil {
				panic(err)
			}
			wire = append(wire, Segment{string(part[0]), n})
		}
		wires[i] = wire
	}
	return wires
}

func solve(puzzleInput [2][]Segment) (int, int) {
	wire1Map := make(map[[2]int]int)
	pos := [2]int{0, 0}
	wire1Map[pos] = 0
	d := 1

	for _, segment := range puzzleInput[0] {
		direction := directions[segment.direction]
		for range segment.distance {
			pos[0] += direction[0]
			pos[1] += direction[1]
			if _, ok := wire1Map[pos]; !ok {
				wire1Map[pos] = d
			}
			d++
		}
	}

	pos = [2]int{0, 0}
	wire2Map := make(map[[2]int]int)
	wire2Map[pos] = 0
	d = 1

	var intersections [][2]int
	for _, segment := range puzzleInput[1] {
		direction := directions[segment.direction]
		for range segment.distance {
			pos[0] += direction[0]
			pos[1] += direction[1]
			if _, ok := wire1Map[pos]; ok {
				intersections = append(intersections, pos)
			}
			if _, ok := wire2Map[pos]; !ok {
				wire2Map[pos] = d
			}
			d++
		}
	}

	part1 := math.MaxInt
	part2 := math.MaxInt
	for _, pos := range intersections {
		manDist := utils.AbsInt(pos[0]) + utils.AbsInt(pos[1])
		if manDist < part1 {
			part1 = manDist
		}
		steps := wire1Map[pos] + wire2Map[pos]
		if steps < part2 {
			part2 = steps
		}
	}

	return part1, part2
}
